pycheribuild/projects/adduser.py

```diff
@@ -0,0 +1,91 @@
+#
+# THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
+# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+# OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+# HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+# OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+# SUCH DAMAGE.
+#
+
+import os
+import getpass
+import subprocess
+import sys
+
+from pathlib import Path
+
+from .project import CheriConfig, SimpleProject, DefaultInstallDir
+from ..utils import (AnsiColour, coloured, status_update)
+from ..processutils import run_command
+
+
+class AddUser(SimpleProject):
+    target = "adduser"
+    native_install_dir = DefaultInstallDir.BOOTSTRAP_TOOLS
+
+    def __init__(self, config: CheriConfig):
+        super().__init__(config)
+
+    def process(self):
+        file = str(self.config.output_root.absolute()) + "/Dockerfile.adduser"
+        try:
+            user = getpass.getuser()
+        except KeyError:
+            user = "nobody"
+
+        # Create a Dockerfile that will contain this user's name, gid, uid
+        try:
+            fp = open(file, mode="w")
+            contents = "FROM cheribuild-docker\n\nRUN addgroup --gid " + \
+                       str(os.getgid()) + " " + user + \
+                       " && adduser --uid " + str(os.getuid()) + " --ingroup " + \
+                       user + " " + user + " \n"
+
+            fp.write(contents)
+
+            fp.close()
+
+        except OSError:
+            status_update("Could not create ", file)
+            sys.exit()
+
+        # Build a new image from our installed image with this user
+        try:
+            docker_run_cmd = ["docker", "build", "--tag=cheribuild-docker",
+                              "-f", file, "."]
+            run_command(docker_run_cmd, config=self.config)
+
+            os.remove(file)
+
+        except subprocess.CalledProcessError as e:
+            os.remove(file)
+
+            # if the image is missing print a helpful error message:
+            if e.returncode == 125:
+                cheribuild_dir = str(Path(__file__).absolute().parent.parent)
+                status_update("It seems like the docker image", self.config.docker_container, "was not found.")
+                status_update("In order to build the default docker image for cheribuild (cheribuild-docker) run:")
+                print(coloured(AnsiColour.blue, "cd", cheribuild_dir +
+                               "/docker && docker build --tag cheribuild-docker ."))
```

Comment:
Wouldn't `docker build --tag cheribuild-docker /path/to/cheribuild/docker` also work, no cd needed? But see above, not sure this should be reported to the user.

Reply:
These are the exact directlons from main.

Reply:
As above. And copying code is a good time to review whether the original code from many years ago that was hacked together and lightly tested is the best way of doing it.